### PR TITLE
[SPARK-8743] [Streaming]: Deregister Codahale metrics for streaming when StreamingContext is closed

### PR DESCRIPTION
**Pull request by nssalian**
The issue link: https://issues.apache.org/jira/browse/SPARK-8743
Deregister Codahale metrics for streaming when StreamingContext is closed

Design:
Adding the method calls in the appropriate start() and stop () methods for the StreamingContext

Actions in the PullRequest:
1) Added the registerSource method call to the start method for the Streaming Context. 
2) Added the removeSource method to the stop method. 
3) Added comments for both 1 and 2 and comment to show initialization of the StreamingSource
4) Added a test case to check for both registration and de-registration of metrics

Previous closed PR for reference: https://github.com/apache/spark/pull/7250

### CODE REVIEW

**Comment by nssalian** [comment]
@tdas and @srowen, the new PR for the JIRA.

Thank you.

File: streaming/src/main/scala/org/apache/spark/streaming/StreamingContext.scala [hunk 1]
```diff
@@ -192,11 +192,8 @@ class StreamingContext private[streaming] (
       None
     }
 
-  /** Register streaming source to metrics system */
+  /* Initializing a streamingSource to register metrics */
```
Comment (srowen) [inline streaming/src/main/scala/org/apache/spark/streaming/StreamingContext.scala]: Not a big deal, but I don't see why this was changed

File: streaming/src/test/scala/org/apache/spark/streaming/StreamingContextSuite.scala [hunk 2]
```diff
@@ -33,8 +33,12 @@ import org.apache.spark.storage.StorageLevel
 import org.apache.spark.streaming.dstream.DStream
 import org.apache.spark.streaming.receiver.Receiver
 import org.apache.spark.util.Utils
-import org.apache.spark.{Logging, SparkConf, SparkContext, SparkException, SparkFunSuite}
-
+import org.apache.spark.{Logging, SparkConf, SparkContext, SparkFunSuite}
```
Comment (srowen) [inline streaming/src/test/scala/org/apache/spark/streaming/StreamingContextSuite.scala]: Again, your imports are not organized properly. Please pay careful attention to this. spark imports go together and are sorted alphabetically. scala imports should go together too.

File: streaming/src/main/scala/org/apache/spark/streaming/StreamingContext.scala (hunk 1 above)
Comment (nssalian) [inline streaming/src/main/scala/org/apache/spark/streaming/StreamingContext.scala]: It previously held the block of code that did the registration.
Here it simply initializes the streamingSource

File: streaming/src/test/scala/org/apache/spark/streaming/StreamingContextSuite.scala (hunk 2 above)
Comment (nssalian) [inline streaming/src/test/scala/org/apache/spark/streaming/StreamingContextSuite.scala]: Made the change in the next commit.

**Comment by nssalian** [comment]
Added the changes and ran ./dev/scalastyle and ~test-only *StreamingContextSuite.
Both passed.

**Comment by nssalian** [comment]
@tdas does this PR need more improvement?

File: streaming/src/test/scala/org/apache/spark/streaming/StreamingContextSuite.scala [hunk 3]
```diff
@@ -299,6 +302,26 @@ class StreamingContextSuite extends SparkFunSuite with BeforeAndAfter with Timeo
...
+    ssc = new StreamingContext(conf, batchDuration)
+    assert(ssc.getState() === StreamingContextState.INITIALIZED)
+    addInputStream(ssc).register()
+    ssc.start()
+
+    val sources = StreamingContextSuite.getSources(ssc.env.metricsSystem)
+    val streamingSource = StreamingContextSuite.getStreamingSource(ssc)
+    assert(sources.contains(streamingSource))
+    assert(ssc.getState() === StreamingContextState.ACTIVE)
+    Thread.sleep(100)
```
Comment (tdas) [inline streaming/src/test/scala/org/apache/spark/streaming/StreamingContextSuite.scala]: Any particular reason why you added this sleep?

**Comment by tdas** [comment]
LGTM, minus a nit on the Thread.sleep().

File: streaming/src/test/scala/org/apache/spark/streaming/StreamingContextSuite.scala (hunk 3 above)
Comment (nssalian) [inline streaming/src/test/scala/org/apache/spark/streaming/StreamingContextSuite.scala]: Removed it. Added it during my runs. 

Thanks.

**Comment by nssalian** [comment]
@tdas Removed it.
Thank you for the review.

**Comment by tdas** [comment]
Jenkins, this is ok to test.

**Comment by andrewor14** [comment]
Jenkins, slow test please

**Comment by tdas** [comment]
LGTM! Merging this in master and branch 1.4

**Comment by tdas** [comment]
This has been merged. But for some reason the PR was not automatically closed. So could you close this PR. Thanks!

**Comment by nssalian** [comment]
Closing. Thank you.

**Pull request closed without merging**